docs/proposals/petset.md

```diff
@@ -0,0 +1,533 @@
+{
+    "spec": {
+        // ...
+
+        "supportTemplates": [{
+            "persistentVolumeClaim": {
+                "spec": {...}
+            },
+            "transform": [{"op": "add", "path": "/metadata/name", "value": "pod-name-pvc-$(IDENTITY)"}]
+        }]
+    }
+}
+
+```
+
+Of all proposals, 2e is the closest to actual semantics we support today (variable replacement in EnvVar and
+patch operations). The path syntax from RFC6902 is not optimal (we prefer JQ field path semantics), but we
+could consider supporting both. Since we desire to support an integer value for quorum size, we may need to
+transform the set size value to an integer for replacement into a supporting resource that needs that value
+as a number:
+
+```
+// Proposal 2f: variant: specify merge-patch semantics with some variable replacement as an integer
+{
+    "spec": {
+        // ...
+
+        "supportTemplates": [{
+            "replicationController": {
+                "spec": {"replicas": 1}
+            },
+            // "$(REPLICAS)" is transformed into 5, not "5"
+            "transform": [{"op": "add", "path": "/metadata/name", "value": "$(REPLICAS)", type: "number"}]
+        }]
+    }
+}
+
+```
+
+
+
+### Accessing pods by stable network identity
+
+In order to provide stable network identity, given that pods may not assume pod IP is constant over the
+lifetime of a pod, it must be possible to have a resolvable DNS name for the pod that is tied to the
+pod identity. There are two broad classes of clustered services - those that require clients to know
+all members of the cluster (load balancer intolerant) and those that are amenable to load balancing.
+For the former, clients must also be able to easily enumerate the list of DNS names that represent the
+member identities and access them inside the cluster. Within a pod, it must be possible for containers
+to find and access that DNS name for identifying itself to the cluster.
+
+Since a pod is expected to be controlled by a single controller at a time, it is reasonable for a pod to
+have a single identity at a time. Therefore, a service can expose a pod by its identity in a unique
+fashion via DNS by leveraging information written to the endpoints by the endpoints controller. The
+mechanism for that is either generic metadata delivered to the pod by the PetSet controller (an identity
+field as in 1c, or an annotation as in 1b).
+
+The end result might be DNS resolution as follows:
+
+```
+# service mongo pointing to pods created by PetSet mdb, with identities mdb-1, mdb-2, mdb-3
+
+dig mongodb.namespace.svc.cluster.local +short A
+172.130.16.50
+
+dig mdb-1.mongodb.namespace.svc.cluster.local +short A
+# IP of pod created for mdb-1
+
+dig mdb-2.mongodb.namespace.svc.cluster.local +short A
+# IP of pod created for mdb-2
+
+dig mdb-3.mongodb.namespace.svc.cluster.local +short A
+# IP of pod created for mdb-3
+
+# made up name, returns a set of CNAMES? or SRV records for each identity ???
```

Comment:
We have enough info in the endpoints so that the creation of the DNS set also assigns the shard name.  Note this service does not have to be headless - it could be a load balanced endpoint that the client uses to get an instance, and then discovers the membership set from the cluster after joining.